File: Published/SourceCode/Java/Graphs/GraphTraversal.java

static void <V> traverseDFS(Graph<V> G, V v, Set<V> visitedVertices) {
    PreVisit(G, v);
    visitedVertices.add(v);
    for (Edge<V> e : G.outgoingEdges(v)) {
        V w = e.end;
        if (!visitedVertices.contains(w))
            traverseDFS(G, w, visitedVertices);
    }
    PostVisit(G, v);
}

static void <V> traverseBFS(Graph<V> G, V v, Set<V> visitedVertices) {
    Queue<V> Q = new Queue<>();
    Q.enqueue(v);
    visitedVertices.add(v);
    while (!Q.isEmpty()) { // Process each vertex on Q
        V v = Q.dequeue();
        PreVisit(G, v);
        for (Edge<V> e : G.outgoingEdges(v)) {
            V w = e.end;
            if (!visitedVertices.contains(w)) { // Put neighbors on Q
                visitedVertices.add(w);
                Q.enqueue(w);
            }
        }
        PostVisit(G, v);
    }
}

static void <V> graphTraverse(Graph<V> G) {
    Set<V> visitedVertices = new Set<V>();
    for (V v : G.vertices())
        if (!visitedVertices.contains(v))
            doTraversal(G, v, visitedVertices);

File: Published/SourceCode/Python/Graphs/GraphTraversal.py

from API import Graph, Edge, Set

def traverseDFS(G, v, visited):
    if v not in visited:
        visited.add(v)
        PreVisit(G, v)
        for edge in G.outgoingEdges(v):
            traverseDFS(G, edge.end, visited)
        PostVisit(G, v)


def traverseDFS(G, v, visited):
    agenda = Stack()
    agenda.push(v)
    while not agenda.isEmpty()
        v = agenda.pop()
        if not visited.contains(v):
            visited.add(v)
            PreVisit(G, v)
            for edge in G.outgoingEdges(v):
                agenda.push(edge.end)
            # PostVisit is not possible in a stack/queue based version!


def traverseBFS(G, v, visited):
    agenda = Queue()
    agenda.enqueue(v)
    while not agenda.isEmpty()
        v = agenda.dequeue()
        if not visited.contains(v):
            visited.add(v)
            PreVisit(G, v)
            for edge in G.outgoingEdges(v):
                agenda.enqueue(edge.end)
            # PostVisit is not possible in a stack/queue based version!


def graphTraverse(G):
    Visited = Set()
    for each vertex v in G.vertices():
        if not Visited.contains(v):
            doTraversal(G, v, Visited)
